### ISSUE TITLE
Time formating in parser seems broken

### ISSUE THREAD

**Issue by chkt** [fluent/fluentd#2512]
### The problem
I am trying to reformat the time key of nginx logs on the fly using the `types` keyword. None of the formats proposed by the documentation work as expected and most format the time key to `null`.

### Your version
I am using fluentd **1.4.2** inside the official docker container

### Your configuration
```ini
<source>
  @type tail
 
  tag nginx
  path /var/log/www_access.log

  <parse>
    @type syslog
  </parse>
</source>

<match nginx>
  @type rewrite_tag_filter

  <rule>
    key message
    pattern /^(\d{1,3}\.){3}\d{1,3} - /
    tag nginx_access
  </rule>
</match>

<filter nginx_access>
  @type parser

  key_name message
  
  <parse>
    @type regexp
  
    expression /^(?<remote>[^ ]*) (?<host>[^ ]*) (?<user>[^ ]*) \[(?<access_time>[^\]]*)\] "(?:(?<method>\S+)(?: +(?<path>[^\"]*?)(?: +(?<protocol>HTTP\/\d\.\d))?)?)?" (?<code>[^ ]*) (?<size>[^ ]*)(?: "(?<referer>[^\"]*)" "(?<agent>[^\"]*)"(?:\s+(?<http_x_forwarded_for>[^ ]+))?)?$/
    types access_time:time:unixtime
  </parse>
</filter>

<match nginx_access>
  @type file

  path /var/log/www_test.log
  append true

  <format>
    @type json
  </format>
</match>
```

### Expected behavior
A log line like

 `127.0.0.1 - - [14/Jul/2019:04:33:36 +0000] "GET / HTTP/1.1" 200 1234 "-" "curl/7.47.0"` 

should be parsed into an entry like 

`{"remote":"127.0.0.1", "access_time":"1563078816", "method":"GET", "path":"/","protocol":"HTTP/1.1", "code":"200", "size":"1234", "referer":"-", "agent":"curl/7.47.0"}`

### Actual behavior
The log line will be parsed to `{"remote":"127.0.0.1", "access_time":"14", "method":"GET", "path":"/","protocol":"HTTP/1.1", "code":"200", "size":"1234", "referer":"-", "agent":"curl/7.47.0"}`

Note that this is merely an example. I also tried other formats like ISO8601 and none other than leaving the original value alone worked. When using something like `access_time:time:%Y%m%D` the result inevitable become `null`.

**Comment by repeatedly** [comment]
Hmm... your log line example is not syslog.
Your tail conf can parse this log?

**Comment by chkt** [comment]
Well, I simplified a little bit for the sake of the example. The logs are actually from a docker container logging to systemd, which has been set up on the host side to forward to syslogd; and in the real world the file is not a separate access log, but messages.log. This setup is obviously not so great, and is in the process of being changed. But since I want to retain the logs that already exist, I set up an additional pipeline to parse the old logs into the new setup.

I already designed all the logic necessary to filter the access logs from other messages, but that is a whole lot of configuration that is imho unrelated to the problem encountered, especially since I encounter the same problem when it try to reformat the access time in my production pipeline. If you think that the behavior is caused by another problem of my setup, I can also post the unabbreviated version of the config.

**Comment by flowl** [comment]
@chkt Did you solve this?
I have the same problem with 1.7, none of the stuff regarding time formatting seems to work.
https://github.com/fluent/fluentd/issues/2718

**Comment by flowl** [comment]
@chkt 

For other users seeking for help:

`types` seems broken or I don't understand it. Use the record transformer instead to build a proper ES timestamp (which can be used by Kibana to build a datetime axis by default <fireworks.jpeg>).

<filter apache.**>
  @type record_transformer
  enable_ruby
  <record>
    # time is the object representation of your datetime var (in this case "@timestamp").
    @timestamp ${time.strftime('%Y-%m-%dT%H:%M:%S%z')}
  </record>
</filter>

**Comment by chkt** [comment]
@flowl No, I have not. For me the solution was to also use a record_transformer in a way similar to your example. I do not know how much of a performance hit this is compared to formatting the date right in the parser, but it certainly is a whole lot less self-explanatory…

**Comment by onet-git** [comment]
It is confusing because `types` is documented to indicate what the output data type of a field should be, yet in the case of `time` is instead specifies the input data format!

`types access_time:time:unixtime` tells the parser to parse the access_time field as if it was a time stored in unixtime format (i.e. seconds since epoch).

`types access_time:time:%Y-%m-%dT%H:%M:%S%z` is likely what you actually want.

**Comment by chkt** [comment]
@onet-git yes, that is very counter intuitive. If this is the expected behavior, then the documentation should at least clearly state so.

But the implementation seems to also implicitly fall back to exporting the time in same format as it imported it, and there is no straightforward way to reformat fields in an output plugin either.

If this key is supposed to work this way, I don't even know what the point of it is. Reformatting the output format of the time key is the single most useful type transformation by a long stretch, and if that can only be done using a record_transformer with ruby enabled in the way @flowl documented, then this key might as well not exist at all.

Are we missing something really obvious? Is there any better way to convert a imported date into a different output format?

**Comment by flowl** [comment]
I have played arround with the `types` and found out they apply to how elastic search should be using the value, it wasn't working for me tho: https://github.com/fluent/fluentd/issues/2718

**Comment by repeatedly** [comment]
What does `it wasn't working` mean?

- Parsing is failed
- Parsing is succeeded but sending to ES is failed.

If latter, it seems es plugin configuration issue.

**Comment by flowl** [comment]
> What does `it wasn't working` mean?
> 
> * Parsing is failed
> * Parsing is succeeded but sending to ES is failed.
> 
> If latter, it seems es plugin configuration issue.

First of all, it's a documentation issue. It is not clear what each of the date/time related properties exactly do.
My specific problem was, that the time format sent to ElasticSearch was not recognized properly by Kibana until I used a formatter.

`types access_time:time:%Y-%m-%dT%H:%M:%S%z` somethign like this did not do the trick, although the format I used was exactly the one ES/Kibana expect.
